runtime/src/accounts_hash.rs

```diff
@@ -667,29 +668,43 @@ impl AccountsHash {
         let mut sum: u128 = 0;
         if len > 0 {
             let mut i = 0;
+            let mut insert_item = false;
             // look_for_first_key means the first key we find in our slice may be a
             //  continuation of accounts belonging to a key that started in the last slice.
             // so, look_for_first_key=true means we have to find the first key different than
             //  the first key we encounter in our slice. Note that if this is true,
             //  our slice begins one index prior to the 'actual' start of our logical range.
-            let mut look_for_first_key = !is_first_slice;
             'outer: loop {
                 // at start of loop, item at 'i' is the first entry for a given pubkey - unless look_for_first
-                let now = &slice[i];
-                let last = now.pubkey;
-                if !look_for_first_key && now.lamports != ZERO_RAW_LAMPORTS_SENTINEL {
-                    // first entry for this key that starts in our slice
-                    result.push(now.hash);
-                    sum += now.lamports as u128;
+                let mut now = &slice[i];
```

Comment:
The main change here is that we need to find the last entry for each pubkey. The previous code wanted the first item. Since we read directly from appendvecs and write_versions increase as we read, this is the natural sort order.